File: crates/burn-seedvc/src/wavenet.rs
//! The WaveNet block the diffusion transformer ends with.
//!
//! Eight gated, weight-normalised convolutions over the transformer's output,
//! conditioned on a second embedding of the flow time. It is the same
//! `WN`/`fused_add_tanh_sigmoid_multiply` shape VITS uses — and which
//! [`burn_vits::Wn`] already implements — but **it is not the same module**, for
//! two reasons that each change the arithmetic:
//!
//! - VITS conditions once, adding a single `g` broadcast over every layer. This
//!   one slices a per-layer window out of `cond_layer`'s `2 · hidden · n_layers`
//!   channels, which is what the checkpoint's `[8192, 512, 1]` records
//!   (`2 × 512 × 8`).
//! - The convolutions are upstream's `SConv1d`, which **ignores the `padding`
//!   argument it is handed and reflect-pads instead** — see [`reflect_pad`].
//!
//! Reusing [`burn_vits::Wn`] would therefore have meant bending a module that
//! RVC and GPT-SoVITS both depend on to fit a third caller agreeing with
//! neither. What *is* reused is [`burn_vits::WeightNormConv1d`], which mirrors
//! `torch.nn.utils.weight_norm`'s `weight_g`/`weight_v` pair exactly and needs no
//! change at all.
//!
//! # What is deliberately absent
//!
//! Upstream multiplies by `x_mask` in the residual and again at the output, and
//! drops out the gate at `p_dropout: 0.2`. Neither is modelled: dropout is a
//! training-time term, and the mask is all ones for a single clip — the only
//! batch shape inference builds, the classifier-free-guidance pair being the same
//! clip twice. **A padded batched trainer would have to add the mask back**, or
//! short samples would leak conditioning past their own end.
//!
//! # Provenance
//!
//! Ported from `modules/wavenet.py` and `modules/encodec.py` of Seed-VC
//! (<https://github.com/Plachtaa/seed-vc>, GPL-3.0), read as a reference and
//! never run or vendored.

use burn::module::Module;
use burn::tensor::Tensor;
use burn::tensor::activation::{sigmoid, tanh};
use burn::tensor::backend::Backend;
use burn_vits::WeightNormConv1d;

use crate::config::SeedVcConfig;

/// Pad `x` by reflection, `pad` frames at each end — what upstream's `SConv1d`
/// does instead of the zero padding its call site appears to ask for.
///
/// `WN` builds each layer as `SConv1d(…, padding=(k·d − d)/2, …)`, but
/// `SConv1d.__init__` never forwards `padding` to the `nn.Conv1d` it wraps: the
/// argument disappears into `**kwargs` and the convolution is built unpadded.
/// `SConv1d.forward` then pads the *input* by `k_eff − stride`, split as evenly
/// as an odd total allows, in `reflect` mode. **Reading the call site rather
/// than the wrapper is how this port would have acquired zero padding**, which
/// is not an error anywhere — it simply pulls the first and last `pad` frames of
/// every layer towards silence, eight times over.
///
/// Upstream additionally zero-extends a signal shorter than the padding before
/// reflecting. That path is not modelled: at the mel frame rate two frames is
/// 23 ms, and asserting beats quietly computing something else.
fn reflect_pad<B: Backend>(x: Tensor<B, 3>, pad: usize) -> Tensor<B, 3> {
    if pad == 0 {
        return x;
    }
    let [b, c, frames] = x.dims();
    assert!(
        frames > pad,
        "reflect padding of {pad} needs more than {pad} frames, got {frames}"
    );
    // Reflection excludes the edge sample itself: for pad = 2 the left side is
    // `x[2], x[1]` and the right side is `x[T-3], x[T-4]`.
    let left = x.clone().slice([0..b, 0..c, 1..pad + 1]).flip([2]);
    let right = x
        .clone()
        .slice([0..b, 0..c, (frames - 1 - pad)..(frames - 1)])
        .flip([2]);
    Tensor::cat(vec![left, x, right], 2)
}

/// Seed-VC's `WN`: gated convolutions with per-layer conditioning.
#[derive(Module, Debug)]
pub struct WaveNet<B: Backend> {
    in_layers: Vec<WeightNormConv1d<B>>,
    res_skip_layers: Vec<WeightNormConv1d<B>>,
    cond_layer: WeightNormConv1d<B>,
    hidden: usize,
    pad: usize,
}

impl<B: Backend> WaveNet<B> {
    /// Build the block from the preset's `wavenet_*` dimensions.
    ///
    /// **`dilation_rate` is 1 on this preset**, so `dilation_rate.pow(i)` is 1 at
    /// every layer and the receptive field grows linearly rather than
    /// exponentially — worth knowing before assuming a WaveNet-shaped module has
    /// WaveNet's usual reach. Nothing else here would need changing for a
    /// dilating preset except the per-layer padding, so the constructor refuses
    /// one rather than computing the wrong padding for it.
    pub fn new(cfg: &SeedVcConfig, device: &B::Device) -> Self {
        let hidden = cfg.hidden_dim;
        let n = cfg.wavenet_layers;
        let k = cfg.wavenet_kernel;
        assert_eq!(
            cfg.wavenet_dilation, 1,
            "only dilation_rate 1 is ported; a dilating preset needs per-layer padding"
        );

        Self {
            in_layers: (0..n)
                .map(|_| WeightNormConv1d::new(hidden, 2 * hidden, k, 1, 0, 1, device))
                .collect(),
            // The last layer feeds only the skip sum, so it needs no residual
            // half — which is why `res_skip_layers.7` is `[512, …]` where the
            // other seven are `[1024, …]`.
            res_skip_layers: (0..n)
                .map(|i| {
                    let out = if i + 1 < n { 2 * hidden } else { hidden };
                    WeightNormConv1d::new(hidden, out, 1, 1, 0, 1, device)
                })
                .collect(),
            cond_layer: WeightNormConv1d::new(hidden, 2 * hidden * n, 1, 1, 0, 1, device),
            hidden,
            pad: (k - 1) / 2,
        }
    }

    /// `x`: `[batch, hidden, frames]`, `g`: `[batch, hidden, 1]` — the flow-time
    /// embedding. Returns `[batch, hidden, frames]`.
    ///
    /// `g` is projected once and then read a window at a time, so each layer sees
    /// a different 1024-channel slice of the same projection.
    pub fn forward(&self, x: Tensor<B, 3>, g: Tensor<B, 3>) -> Tensor<B, 3> {
        let g = self.cond_layer.forward(g);
        let n = self.in_layers.len();

        let mut x = x;
        let mut output = Tensor::zeros_like(&x);
        for i in 0..n {
            let acts = self.in_layers[i].forward(reflect_pad(x.clone(), self.pad))
                + g.clone().narrow(1, i * 2 * self.hidden, 2 * self.hidden);
            // `fused_add_tanh_sigmoid_multiply`: the first half is the value, the
            // second the gate.
            let acts = tanh(acts.clone().narrow(1, 0, self.hidden))
                * sigmoid(acts.narrow(1, self.hidden, self.hidden));

            let res_skip = self.res_skip_layers[i].forward(acts);
            if i + 1 < n {
                x = x + res_skip.clone().narrow(1, 0, self.hidden);
                output = output + res_skip.narrow(1, self.hidden, self.hidden);
            } else {
                output = output + res_skip;
            }
        }
        output
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    type B = burn_ndarray::NdArray;

    #[test]
    fn reflection_mirrors_the_edges_without_repeating_them() {
        // The one piece of arithmetic here a reader would get wrong from the call
        // site alone, so it is pinned rather than inferred.
        let device = Default::default();
        let x = Tensor::<B, 1>::from_floats([1.0, 2.0, 3.0, 4.0, 5.0], &device).reshape([1, 1, 5]);
        let padded: Vec<f32> = reflect_pad(x, 2).into_data().to_vec().unwrap();
        assert_eq!(padded, vec![3.0, 2.0, 1.0, 2.0, 3.0, 4.0, 5.0, 4.0, 3.0]);
    }

    #[test]
    fn the_block_is_length_preserving_and_finite() {
        let cfg = SeedVcConfig::uvit_whisper_small_wavenet();
        let device = Default::default();
        let net = WaveNet::<B>::new(&cfg, &device);

        let x = Tensor::zeros([1, cfg.hidden_dim, 24], &device);
        let g = Tensor::zeros([1, cfg.hidden_dim, 1], &device);
        let out = net.forward(x, g);
        assert_eq!(out.dims(), [1, cfg.hidden_dim, 24]);
        // Separate from any value check: Burn's approximate comparisons treat NaN
        // as equal to NaN, so a NaN would pass one silently.
        assert!(!out.contains_nan().into_scalar());
    }
}

File: export/seedvc_infer/wavenet.py
"""Standalone torch implementation of Seed-VC's WaveNet tail.

Mirrors `crates/burn-seedvc/src/wavenet.rs` field for field, so the released
checkpoint loads onto it by direct key mapping. This is a clean-room
reimplementation written from the Burn port: it does not import, vendor or run
the Seed-VC repository.

The block has no graph of its own — it is the second half of `dit.onnx` — so
there is no `Graph` class here. What it does own is the one piece of arithmetic
a reader cannot get from the call site: **the convolutions reflect-pad**, which
is upstream's `SConv1d` ignoring the `padding` argument it is handed. See
[`reflect_pad`].

Weight-norm is **not** folded, unlike `gptsovits_infer.py`. The exporter loads
the checkpoint's `weight_g`/`weight_v` pairs straight onto these parameters, so
folding here would mean the mirror's parameter names stopped matching the file
it is fed.
"""

from __future__ import annotations

import torch
import torch.nn.functional as F
from torch import nn


# ---- weight norm ------------------------------------------------------------


class WeightNormConv1d(nn.Module):
    """`torch.nn.utils.weight_norm(nn.Conv1d)`, mirroring `burn_vits::WeightNormConv1d`.

    `weight_g` is `[out, 1, 1]` and `weight_v` is `[out, in, kernel]` — weight
    norm's default `dim=0`, one magnitude per output channel. The norm therefore
    runs over the input and kernel axes and nothing else.
    """

    def __init__(self, in_ch: int, out_ch: int, kernel: int) -> None:
        super().__init__()
        self.weight_g = nn.Parameter(torch.ones(out_ch, 1, 1))
        self.weight_v = nn.Parameter(torch.randn(out_ch, in_ch, kernel) * 0.02)
        self.bias = nn.Parameter(torch.zeros(out_ch))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        v = self.weight_v
        weight = v * (self.weight_g / v.pow(2).sum(dim=(1, 2), keepdim=True).sqrt())
        return F.conv1d(x, weight, self.bias)


# ---- the block --------------------------------------------------------------


def reflect_pad(x: torch.Tensor, pad: int) -> torch.Tensor:
    """Pad `x` by reflection, `pad` frames at each end.

    `WN` builds each layer as `SConv1d(…, padding=(k·d − d)/2, …)`, but
    `SConv1d.__init__` never forwards `padding` to the `nn.Conv1d` it wraps: the
    argument disappears into `**kwargs` and the convolution is built unpadded.
    `SConv1d.forward` then reflect-pads the *input* instead. **Reading the call
    site rather than the wrapper is how this mirror would have acquired zero
    padding**, which is not an error anywhere — it simply pulls the first and
    last `pad` frames of every layer towards silence, eight times over.
    """
    return x if pad == 0 else F.pad(x, (pad, pad), mode="reflect")


class WaveNet(nn.Module):
    """Seed-VC's `WN`: gated convolutions with **per-layer** conditioning.

    Not VITS's `WN`, which adds one broadcast `g` to every layer. This one
    projects `g` once into `2 · hidden · layers` channels and reads a different
    window of it per layer — which is what the checkpoint's `cond_layer` weight
    of `[8192, 512, 1]` records, `2 × 512 × 8`.

    Upstream's `x_mask` and its `p_dropout: 0.2` are both absent: dropout is a
    training-time term, and the mask is all ones for the single clip inference
    ever passes (the classifier-free-guidance pair being that clip twice).
    """

    def __init__(self, hidden: int, layers: int, kernel: int, dilation: int) -> None:
        super().__init__()
        if dilation != 1:
            raise ValueError("only dilation_rate 1 is mirrored; a dilating preset needs per-layer padding")
        self.hidden = hidden
        self.pad = (kernel - 1) // 2
        self.in_layers = nn.ModuleList(
            WeightNormConv1d(hidden, 2 * hidden, kernel) for _ in range(layers)
        )
        # The last layer feeds only the skip sum, so it needs no residual half —
        # which is why `res_skip_layers.7` is `[512, …]` where the other seven
        # are `[1024, …]`.
        self.res_skip_layers = nn.ModuleList(
            WeightNormConv1d(hidden, 2 * hidden if i + 1 < layers else hidden, 1)
            for i in range(layers)
        )
        self.cond_layer = WeightNormConv1d(hidden, 2 * hidden * layers, 1)

    def forward(self, x: torch.Tensor, g: torch.Tensor) -> torch.Tensor:
        """`x`: `[batch, hidden, frames]`, `g`: `[batch, hidden, 1]` → `[batch, hidden, frames]`."""
        g = self.cond_layer(g)
        n, h = len(self.in_layers), self.hidden

        output = torch.zeros_like(x)
        for i in range(n):
            acts = self.in_layers[i](reflect_pad(x, self.pad)) + g[:, i * 2 * h : (i + 1) * 2 * h]
            # `fused_add_tanh_sigmoid_multiply`: first half the value, second the gate.
            acts = torch.tanh(acts[:, :h]) * torch.sigmoid(acts[:, h:])

            res_skip = self.res_skip_layers[i](acts)
            if i + 1 < n:
                x = x + res_skip[:, :h]
                output = output + res_skip[:, h:]
            else:
                output = output + res_skip
        return output
